rfd/0035-aws-node-join.md

```diff
@@ -0,0 +1,261 @@
+   requested Teleport service role.
+4. Check that this EC2 instance has not already joined the cluster.
+   - The node name will be set to `<aws_account_id>-<aws_instance_id>` so that
+     the auth server can efficiently check the backend to see whether this
+     instance has already joined.
+   - This is to prevent replay attacks with the IID
+5. Check that the EC2 instance is currently running. \*
+
+\* Step 5 requires AWS credentials on the Auth server with permissions for
+`ec2:DescribeInstances`. If you need Nodes to be able to join the teleport
+cluster from different AWS accounts than the one in which the Teleport Auth
+server is running, there are a few extra required steps:
+1. Every AWS account in which Nodes should be able to join will need an IAM
+   role which can be assumed from the account of the Auth server. This role
+   needs permissions for `ec2:DescribeInstances`.
+2. The IAM role of the Auth server needs permissions to `sts:AssumeRole` for
+   the ARN of every foreign IAM role from step 1. These ARNs will need to be
+   configured in the AWS join token.
+
+### IAM Method
+
+In place of a join token, nodes will present a signed `sts:GetCallerIdentity`
+request to the auth server.
+
+The form of this request is an HTTP `POST` request to `sts.amazonaws.com` with
+`Action=GetCallerIdentity` in the body. The request can include arbitrary
+headers which will also be signed using the
+[Signature Version 4 signing process](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html).
+The signature will then be included in the `Authorization` header of the request.
+See Appendix I for an example.
+
+The actual creation and signing of the request will be done on the node with
+the AWS SDK for Go.
+
+Because the signed request can include arbitrary headers, this allows us to
+issue a challenge (a crypto random string of bytes) that the node must include
+in the signed request. We can make use of gRPC bidirectional streams to
+implement the following flow at a new `RegisterWithAWSToken` gRPC endpoint:
+
+1. The node initiates a `RegisterWithAWSToken` request.
+2. The auth server enforces that TLS is used for transport.
+3. The auth server sends a base64 encoded 32 byte crypto-random challenge.
+4. The node creates the `sts:GetCallerIdentity` request, adds the challenge in
+   a header, and signs it with its AWS credentials.
+5. The node sends a join request to the auth server, including the signed
+   `sts:GetCallerIdentity` request and other parameters currently in
+   [RegisterUsingTokenRequest](
+   https://github.com/gravitational/teleport/blob/d4247cb150d720be97521347b74bf9c526ae869f/lib/auth/auth.go#L1538-L1563)
+6. The auth server checks that the signed request is valid:
+   - it is a `POST` request to `sts.amazonaws.com`
+   - the body is `Action=GetCallerIdentity&Version=2011-06-15`
+   - the `X-Teleport-Challenge` header value equals the issued challenge
```

Comment:
Excessive length check + secure comparison should be preferred here

Reply:
I missed that only 1 attempt was allowed! If that's the case I wouldn't worry too much, we can keep it as it is.